File: impl/ts/src/domain/hex.ts
// Hex geometry for the Cybernet. Blocks are hexagonal tiles connected edge to
// edge. Mirrors impl/go/internal/domain/hex.go exactly — same direction ordering,
// edge-facing convention, and rotation convention (the parity/determinism
// contract in DOCS/parity.md). See DOCS/domain-model.md ("Cybernet").
//
// Coordinates are axial (q, r). The six edges of a hex are indexed 0..5 and each
// edge i faces grid direction i. HEX_DIRECTIONS[i] is the delta to the neighbor
// across edge i. The neighbor's edge facing back is opposite(i).

export interface Coord {
  q: number;
  r: number;
}

/** Neighbor deltas per edge/grid direction. Shared with the Go mirror; do not reorder. */
export const HEX_DIRECTIONS: readonly Coord[] = [
  { q: 1, r: 0 },
  { q: 1, r: -1 },
  { q: 0, r: -1 },
  { q: -1, r: 0 },
  { q: -1, r: 1 },
  { q: 0, r: 1 },
];

/** Returns the coordinate across edge/direction dir (0..5). */
export function neighbor(c: Coord, dir: number): Coord {
  const d = HEX_DIRECTIONS[(((dir % 6) + 6) % 6)]!;
  return { q: c.q + d.q, r: c.r + d.r };
}

/** Returns the edge that faces back across a shared border. */
export function opposite(dir: number): number {
  return (((dir % 6) + 6 + 3) % 6);
}

/** Stable string key for a coordinate (map keys / equality). */
export function coordKey(c: Coord): string {
  return `${c.q},${c.r}`;
}

export function coordEqual(a: Coord, b: Coord): boolean {
  return a.q === b.q && a.r === b.r;
}

/**
 * A block instance positioned in the Cybernet with a rotation. A block's local
 * edge e faces grid direction (e + rotation) mod 6.
 */
export interface PlacedBlock {
  blockId: string;
  rotation: number;
  coord: Coord;
}

/**
 * The growing hex layout of placed blocks. Blocks are stored in placement order
 * (deterministic iteration); lookups scan the small array.
 */
export class Cybernet {
  blocks: PlacedBlock[] = [];

  at(c: Coord): PlacedBlock | undefined {
    return this.blocks.find((pb) => coordEqual(pb.coord, c));
  }

  occupied(c: Coord): boolean {
    return this.at(c) !== undefined;
  }
}

export function newCybernet(): Cybernet {
  return new Cybernet();
}

File: impl/go/internal/domain/hex.go
package domain

// Hex geometry for the Cybernet. Blocks are hexagonal tiles connected edge to
// edge. This mirrors impl/ts/src/domain/hex.ts exactly — same direction ordering,
// edge-facing convention, and rotation convention (the parity/determinism
// contract in DOCS/parity.md). See DOCS/domain-model.md ("Cybernet").
//
// Coordinates are axial (Q, R). The six edges of a hex are indexed 0..5 and each
// edge i faces grid direction i. HexDirections[i] is the coordinate delta to the
// neighbor across edge i. The neighbor's edge facing back is Opposite(i).

// Coord is an axial hex coordinate.
type Coord struct {
	Q int `json:"q"`
	R int `json:"r"`
}

// HexDirections maps each of the 6 edge/grid directions to a neighbor delta.
// The ordering is canonical and shared with the TS mirror; do not reorder.
var HexDirections = [6]Coord{
	{Q: 1, R: 0},
	{Q: 1, R: -1},
	{Q: 0, R: -1},
	{Q: -1, R: 0},
	{Q: -1, R: 1},
	{Q: 0, R: 1},
}

// Neighbor returns the coordinate across edge/direction dir (0..5).
func (c Coord) Neighbor(dir int) Coord {
	d := HexDirections[((dir%6)+6)%6]
	return Coord{Q: c.Q + d.Q, R: c.R + d.R}
}

// Opposite returns the edge that faces back across a shared border.
func Opposite(dir int) int {
	return ((dir%6)+6+3)%6
}

// PlacedBlock is a block instance positioned in the Cybernet with a rotation.
// A block's local edge e faces grid direction (e + Rotation) mod 6.
type PlacedBlock struct {
	BlockID  string `json:"blockId"`
	Rotation int    `json:"rotation"`
	Coord    Coord  `json:"coord"`
	// OwnerID is the player controlling this block (via Icebreaker), or "" if
	// uncontrolled. A controlled block carries one of that player's control markers.
	OwnerID string `json:"ownerId,omitempty"`
}

// Cybernet is the growing hex layout of placed blocks and the pawns on them.
// Both are stored in insertion order (deterministic iteration); lookups scan the
// small slices.
type Cybernet struct {
	Blocks []*PlacedBlock `json:"blocks"`
	Pawns  []*PawnOnBoard `json:"pawns"`
}

// NewCybernet returns an empty Cybernet.
func NewCybernet() *Cybernet {
	return &Cybernet{Blocks: []*PlacedBlock{}, Pawns: []*PawnOnBoard{}}
}

// At returns the placed block at coord c, or nil if the cell is empty.
func (cy *Cybernet) At(c Coord) *PlacedBlock {
	for _, pb := range cy.Blocks {
		if pb.Coord == c {
			return pb
		}
	}
	return nil
}

// Occupied reports whether cell c holds a block.
func (cy *Cybernet) Occupied(c Coord) bool {
	return cy.At(c) != nil
}

// place adds a block instance (no validation — callers validate first).
func (cy *Cybernet) place(pb *PlacedBlock) {
	cy.Blocks = append(cy.Blocks, pb)
}
